File: Applications/CSharp/Backtracking/CombinationSum2.cs
//  40. Combination Sum II
//  Given a collection of candidate numbers (candidates) and a target number (target), find all unique combinations in candidates where the candidate numbers sum to target.
//  Each number in candidates may only be used once in the combination.
//  Note: The solution set must not contain duplicate combinations.

class Solution
{
    public IList<IList<int>> CombinationSum2(int[] candidates, int target)
    {
        var result = new List<IList<int>>();

        void Dfs(int[] candidates, int idx, int sum, int target, List<int> copy)
        {
            if (sum == target)
            {
                result.Add(copy);
                return;
            }

            if (sum > target || idx >= candidates.Length)
                return;

            Dfs(candidates, idx + 1, sum + candidates[idx], target, new List<int>(copy) { candidates[idx] });

            while (idx + 1 < candidates.Length && candidates[idx] == candidates[idx + 1])
                idx += 1;

            Dfs(candidates, idx + 1, sum, target, new List<int>(copy));
        }

        Dfs(candidates.OrderBy(x => x).ToArray(), 0, 0, target, new List<int>());

        return result;
    }
}

File: Applications/Python/Backtracking/CombinationSum2.py
# 40. Combination Sum II
# Given a collection of candidate numbers (candidates) and a target number (target), find all unique combinations in candidates where the candidate numbers sum to target.
# Each number in candidates may only be used once in the combination.
# Note: The solution set must not contain duplicate combinations.

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def dfs(candidates, idx, target, sum, curLst):
            if sum == target:
                res.append(curLst)
                return            
            
            if sum > target or idx >= len(candidates):
                return
                        
            dfs(candidates, idx + 1, target, sum + candidates[idx], curLst.copy() + [candidates[idx]])
            
            while idx + 1 < len(candidates) and candidates[idx] == candidates[idx + 1]:
                idx += 1

            dfs(candidates, idx + 1, target, sum, curLst.copy())
            
            
        dfs(candidates, 0, target, 0, [])
        return res